File: internal/router/mail.go
package router

import (
	"encoding/json"
	"net/http"
	"strconv"

	"github.com/go-chi/chi/v5"

	"github.com/Neaox/overcast/internal/smtp"
)

// inboxHandlers returns a chi.Router that mounts the inbox capture API under
// /_overcast/inbox/. Only registered when the SMTP mock server is enabled.
//
// Endpoints:
//
//	GET    /_overcast/inbox/messages            list all captured messages (newest first)
//	GET    /_overcast/inbox/messages/{id}       get a single message
//	DELETE /_overcast/inbox/messages            clear all messages
//	DELETE /_overcast/inbox/messages/{id}       delete a single message
func inboxHandlers(store *smtp.MailStore) func(chi.Router) {
	return func(r chi.Router) {
		r.Get("/messages", func(w http.ResponseWriter, req *http.Request) {
			msgs := store.List()

			// Optional ?limit= query param.
			if limitStr := req.URL.Query().Get("limit"); limitStr != "" {
				if n, err := strconv.Atoi(limitStr); err == nil && n > 0 && n < len(msgs) {
					msgs = msgs[:n]
				}
			}

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(msgs) //nolint:errcheck
		})

		r.Get("/messages/{id}", func(w http.ResponseWriter, req *http.Request) {
			id := chi.URLParam(req, "id")
			msg := store.Get(id)
			if msg == nil {
				http.Error(w, `{"error":"not found"}`, http.StatusNotFound)
				return
			}
			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(msg) //nolint:errcheck
		})

		r.Delete("/messages", func(w http.ResponseWriter, req *http.Request) {
			store.Clear()
			w.WriteHeader(http.StatusNoContent)
		})

		r.Delete("/messages/{id}", func(w http.ResponseWriter, req *http.Request) {
			id := chi.URLParam(req, "id")
			if !store.Delete(id) {
				http.Error(w, `{"error":"not found"}`, http.StatusNotFound)
				return
			}
			w.WriteHeader(http.StatusNoContent)
		})
	}
}

File: web/api/src/routes/mail.ts
import { Hono } from "hono"
import { resolveEndpoint, ENDPOINT_HEADER, REGION_HEADER } from "../service-discovery.js"

export const mailRoutes = new Hono()

function endpoint(c: { req: { header: (k: string) => string | undefined } }) {
  return resolveEndpoint({
    [ENDPOINT_HEADER]: c.req.header(ENDPOINT_HEADER),
    [REGION_HEADER]: c.req.header(REGION_HEADER),
  })
}

// ─── List messages ─────────────────────────────────────────────────────────

/**
 * GET /api/mail/messages[?limit=N]
 *
 * Returns all captured SMTP messages, newest first.
 */
mailRoutes.get("/messages", async (c) => {
  const ep = endpoint(c)
  const limit = c.req.query("limit")
  const url = `${ep.baseUrl}/_overcast/inbox/messages${limit ? `?limit=${limit}` : ""}`
  const res = await fetch(url)
  if (!res.ok) return c.json({ error: "mail fetch failed" }, 502)
  return c.json(await res.json())
})

// ─── Get single message ────────────────────────────────────────────────────

/**
 * GET /api/mail/messages/:id
 */
mailRoutes.get("/messages/:id", async (c) => {
  const ep = endpoint(c)
  const res = await fetch(`${ep.baseUrl}/_overcast/inbox/messages/${c.req.param("id")}`)
  if (res.status === 404) return c.json({ error: "not found" }, 404)
  if (!res.ok) return c.json({ error: "mail fetch failed" }, 502)
  return c.json(await res.json())
})

// ─── Clear all messages ────────────────────────────────────────────────────

/**
 * DELETE /api/mail/messages
 *
 * Clears the entire captured mail inbox.
 */
mailRoutes.delete("/messages", async (c) => {
  const ep = endpoint(c)
  const res = await fetch(`${ep.baseUrl}/_overcast/inbox/messages`, { method: "DELETE" })
  if (!res.ok) return c.json({ error: "mail clear failed" }, 502)
  return new Response(null, { status: 204 })
})

// ─── Delete single message ─────────────────────────────────────────────────

/**
 * DELETE /api/mail/messages/:id
 */
mailRoutes.delete("/messages/:id", async (c) => {
  const ep = endpoint(c)
  const res = await fetch(`${ep.baseUrl}/_overcast/inbox/messages/${c.req.param("id")}`, {
    method: "DELETE",
  })
  if (res.status === 404) return c.json({ error: "not found" }, 404)
  if (!res.ok) return c.json({ error: "mail delete failed" }, 502)
  return new Response(null, { status: 204 })
})
